lcls_tools/common/data_analysis/bmad_modeling.py

```diff
@@ -0,0 +1,573 @@
+        self.plot_beta_options = {
+            "title1": "",
+            "title2": "",
+            "label1": "Design",
+            "label2": "Model",
+            "figsize": (8, 4),
+        }
+        self.plot_beta_options.update(kwargs)
+        opt = self.plot_beta_options
+        fig1, ax1 = plt.subplots(figsize=(8, 4))
+        ax1.plot(
+            output1["ele.s"], output1["ele.a.beta"],
+            label=opt["label1"], linestyle="--"
+        )
+        ax1.plot(output2["ele.s"], output2["ele.a.beta"], label=opt["label2"])
+        plt.legend()
+        # Add energy to the rhs
+        ax12 = ax1.twinx()
+        ax12.plot(output2["ele.s"], output2["ele.e_tot"] / 1e9, color="red")
+        ax12.set_ylabel("Energy (GeV)")
+        efinal = output2["ele.e_tot"][-1] / 1e9
+        plt.title(f'{opt["title1"]} Final energy: {efinal:.2f} GeV')
+        ax1.set_xlabel("s (m)")
+        ax1.set_ylabel("Twiss Beta X (m)")
+        # itime = isotime()
+        fig2, ax2 = plt.subplots(figsize=(8, 4))
+        ax2.plot(
+            output1["ele.s"], output1["ele.b.beta"],
+            label=opt["label1"], linestyle="--"
+        )
+        ax2.plot(output2["ele.s"], output2["ele.b.beta"], label=opt["label2"])
+        plt.legend()
+        ax22 = ax2.twinx()
+        ax22.plot(output2["ele.s"], output2["ele.e_tot"] / 1e9, color="red")
+        ax22.set_ylabel("Energy (GeV)")
+        plt.title(f'{opt["title2"]} Final energy: {efinal:.2f} GeV')
+        ax2.set_xlabel("s (m)")
+        ax2.set_ylabel("Twiss Beta Y (m)")
+        axes_list = [ax1, ax12, ax2, ax22]
+        fig1.show()
+        fig2.show()
+        return fig1, fig2, axes_list
+
+    def get_output(self, tao):
+        """Returns dictionary of modeled parameters,
+        including element name, twiss and Rmats"""
+        # Output collecting
+        outkeys = [
+            "ele.name",
+            "ele.ix_ele",
+            "ele.ix_branch",
+            "ele.a.beta",
+            "ele.a.alpha",
+            "ele.a.eta",
+            "ele.a.etap",
+            "ele.a.gamma",
+            "ele.a.phi",
+            "ele.b.beta",
+            "ele.b.alpha",
+            "ele.b.eta",
+            "ele.b.etap",
+            "ele.b.gamma",
+            "ele.b.phi",
+            "ele.x.eta",
+            "ele.x.etap",
+            "ele.y.eta",
+            "ele.y.etap",
+            "ele.s",
+            "ele.l",
+            "ele.e_tot",
+            "ele.p0c",
+            "ele.mat6",
+            "ele.vec0",
+        ]
+        output = {k: tao.lat_list("*", k) for k in outkeys}
+        return output
+
+    def get_element(self, tao, datum):
+        return tao.data_parameter(datum, "ele_name")[0].split(";")[1]
+
+    def update_datum_meas(self, tao, datum, useDesing=True):
+        """Updates datum with EPICS or design twiss values"""
+        element = self.get_element(tao, datum)
+        TWISS = ["beta_a", "alpha_a", "beta_b", "alpha_b"]
```

Comment:
Maybe things like this can go in the yaml too

Reply:
Is this a style question?  Line 400 will never change.